[Content] MemoryRouter reference page (#621)
solidjs/solid-docs-next: src/routes/solid-router/reference/components/memory-router.mdx

---
title: MemoryRouter
---

> The `MemoryRouter` can be used to route while keeping the entire routing history internally in memory.
> Other routers keep the history externally i.e. in the url like the `<HashRouter>` or in the browser history like the [`<Router>`](/solid-router/reference/components/memory-router).

Reviewer: ```suggestion
The `MemoryRouter` can be used to route while keeping the entire routing history within internal memory.
Other routers keep history externally such as in the url like the [`<HashRouter>`](/solid-router/reference/components/hash-router) or in the browser history like [`<Router>`](/solid-router/reference/components/memory-router).
```

Author: I can add this link, just note it'll break CI until the HashRouter page also is merged.

Reviewer: I'll merge them both when they're good to go :)

> Keeping the history in memory is desirable when you want full control over the routing history.
> The MemoryRouter is therefore often used for testing purposes, where it's controlled through manipulating the `MemoryHistory`.

Reviewer: ```suggestion
Keeping the history in memory is useful when you want full control over the routing history.
Since `MemoryRouter` can manipulate `MemoryHistory`, it is often used for testing purposes.
```

> In the example below, a history object pre-filled with a navigation to the `/about` route is passed to the `MemoryRouter`.
> 
> 
> 

Reviewer: ```suggestion
```

```jsx
import { MemoryRouter, createMemoryHistory, A } from "@solidjs/router";
import { Suspense } from "solid-js";

export default function App() {
	const history = createMemoryHistory();

	const toHome = () => {
		history.set({ value: "/" });
	};
	const toAbout = () => {
		history.set({ value: "/about" });
	};

	return (
		<>
			<button onClick={toHome}>{'"/"'}</button>
			<button onClick={toAbout}>{'"/about"'}</button>

			<MemoryRouter
				history={history}
				root={(props) => <Suspense>{props.children}</Suspense>}
			>
				{/*... routes */}
			</MemoryRouter>
		</>
	);
}
```

> 
> ## Manipulating the history

Reviewer: ```suggestion
In this example, a history object is pre-filled to navigate to the `/about` route, which is then passed to the `MemoryRouter`.

## Manipulating the history
```

> The `MemoryHistory` object has the following functions, which are used to control the navigation:

Reviewer: Can you give a quick blurb about what the functions intend to do? Ie. "To give you control over how you navigate your app" or whatever makes the most sense?

```suggestion
The `MemoryHistory` object contains functions that are used to control navigation. 
```

Reviewer: I'm also thinking these are better presented in the form of a table vs a section for each - do you mind putting them into one?

Author: This was a great idea.

I added a table, and some text about each of the three types of functions in the MemoryHistory.
